### ISSUE TITLE
Is it possible to compute different graph on each separate gpu device with Horovod?

### ISSUE THREAD

**Issue by gobigrassland** [horovod/horovod#1652]
For example, I have 10 classification tasks and the feature dimension is 1024. Each classification task is 1000000-way and the weights of the backbone for extracting features are shared. I would like to compute such graphs whose backbone weights are shared but the classification layer are different on each separate gpu device.

**Comment by gobigrassland** [comment]
I try to realize this , but it does not work.
error information:
One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock.

**Comment by tgaddair** [comment]
Hey @gobigrassland, do you have a small example script that illustrates what you're trying to achieve.  I think it should be possible, but just need some clarification about the setup.

In general, as long as the tensors between graphs have unique names or are placed on different devices, it should work.

**Comment by gobigrassland** [comment]
Thank you very much for your reply. The main script which trains a face recognizer with two datasets(Webface and VGGFace2) shows as follows. 

def main(args):
    print(args)
    global IMAGE_HEIGHT
    global IMAGE_WIDTH
    IMAGE_HEIGHT = args.image_height
    IMAGE_WIDTH = args.image_width
    network = importlib.import_module(args.model_def)
    image_size = (args.image_height, args.image_width)
    
    # Initialize Horovod
    hvd.init()

    # Pin GPU to be used to process local rank (one GPU per process)
    config = tf.ConfigProto()
    config.gpu_options.allow_growth = True
    config.gpu_options.visible_device_list = '{}'.format(hvd.local_rank())

    batch_size_1 = 20
    batch_size_2 = 20
    num_threads_1 = 2
    num_threads_2 = 2

    print('Network for training:', args.model_def, args.model_num_layers)
    print('BatchSize:', batch_size_1, batch_size_2)

    # webface       
    data_name_1 = 'webface'
    files_set_1, labels_set_1, num_classes_1 = facenet.get_image_paths_and_labels_beta('', 'train_test_webface.txt')
    files_placeholder_1 = tf.placeholder(tf.string)
    labels_placeholder_1 = tf.placeholder(tf.int32)
    dataset_1 = tf.data.Dataset.from_tensor_slices((files_placeholder_1, labels_placeholder_1))
    dataset_1 = dataset_1.shuffle(len(files_set_1)).repeat(100)
    dataset_1 = dataset_1.prefetch(20)
    dataset_1 = dataset_1.apply(tf.contrib.data.map_and_batch(map_func=parse_function, batch_size=batch_size_1, num_parallel_batches=num_threads_1))
    #iterator_1 = dataset_1.make_initializable_iterator()
    iterator_1 = tf.data.Iterator.from_structure(dataset_1.output_types, dataset_1.output_shapes)
    init_iterator_1 = iterator_1.make_initializer(dataset_1)
    image_batch_1, label_batch_1 = iterator_1.get_next()

    # vggface2
    data_name_2 = 'vggface2'
    files_set_2, labels_set_2, num_classes_2 = facenet.get_image_paths_and_labels_beta('', 'train_test_vgg2.txt')
    files_placeholder_2 = tf.placeholder(tf.string)
    labels_placeholder_2 = tf.placeholder(tf.int32)
    dataset_2 = tf.data.Dataset.from_tensor_slices((files_placeholder_2, labels_placeholder_2))
    dataset_2 = dataset_2.shuffle(len(files_set_2)).repeat(100)
    dataset_2 = dataset_2.prefetch(20)
    dataset_2 = dataset_2.apply(tf.contrib.data.map_and_batch(map_func=parse_function, batch_size=batch_size_2, num_parallel_batches=num_threads_2))
    #iterator_2 = dataset_2.make_initializable_iterator()
    iterator_2 = tf.data.Iterator.from_structure(dataset_2.output_types, dataset_2.output_shapes)
    init_iterator_2 = iterator_2.make_initializer(dataset_2)
    image_batch_2, label_batch_2 = iterator_2.get_next()

    local_rank = hvd.local_rank()
    if local_rank == 0:
        image_batch = image_batch_1
        label_batch = label_batch_1
        num_classes = num_classes_1
        data_name = data_name_1
    else:
        image_batch = image_batch_2
        label_batch = label_batch_2
        num_classes = num_classes_2
        data_name = data_name_2

    # Create the hook to initialize the Iterator with the filenames_list, credit to @ispirmustafa
    class _DatasetInitializerHook(tf.train.SessionRunHook):
        def __init__(self, initializer, feed_dict):
            self._initializer = initializer
            self._feed_dict = feed_dict
        def begin(self):
            pass
        def after_create_session(self, session, coord):
            del coord
            session.run(self._initializer, feed_dict = self._feed_dict)

    prelogits, _ = network.inference(image_batch,
                                     embedding_size=args.embedding_size,
                                     keep_probability=args.keep_probability,
                                     weight_decay=args.weight_decay,
                                     is_training=True if args.is_training == 1 else False,
                                     net_type=args.model_num_layers)
    update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
    if args.debug_info:
        print('UPDATE_OPS: ', len(update_ops), update_ops)
    with tf.control_dependencies(update_ops):
        cross_entropy_mean, accuracy = tower_loss('loss_block', args, prelogits, label_batch, num_classes, data_name)

    regularization_losses = tf.get_collection(tf.GraphKeys.REGULARIZATION_LOSSES)
    cross_entropy_losses = tf.get_collection("losses")
    total_loss = tf.add_n(cross_entropy_losses + regularization_losses, name='total_loss')
    if args.debug_info:
        print(len(cross_entropy_losses), cross_entropy_losses, '\n')
        print(len(regularization_losses), regularization_losses, '\n')

    lr_placeholder = tf.placeholder(tf.float32, name='learning_rate')     
    # Add Horovod Distributed Optimizer
    #opt = tf.train.AdagradOptimizer(lr_placeholder * hvd.size())
    if args.optimizer=='ADAGRAD':
        opt = tf.train.AdagradOptimizer(lr_placeholder * hvd.size())
    elif args.optimizer=='ADADELTA':
        opt = tf.train.AdadeltaOptimizer(lr_placeholder * hvd.size(), rho=0.9, epsilon=1e-6)
    elif args.optimizer=='ADAM':
        opt = tf.train.AdamOptimizer(lr_placeholder * hvd.size(), beta1=0.9, beta2=0.999, epsilon=0.1)
    elif args.optimizer=='RMSPROP':
        opt = tf.train.RMSPropOptimizer(lr_placeholder * hvd.size(), decay=0.9, momentum=0.9, epsilon=1.0)
    elif args.optimizer=='MOM':
        opt = tf.train.MomentumOptimizer(lr_placeholder * hvd.size(), 0.9, use_nesterov=True)
    else:
        raise ValueError('Invalid optimization algorithm')

    exclusions = ['kernel']
    variables_shared = []
    variables_not_shared = []
    for var in tf.global_variables():
      excluded = False
      for exclusion in exclusions:
        #if var.op.name.startswith(exclusion):
        if exclusion in var.op.name:
          excluded = True
          variables_not_shared.append(var)
          break
      if not excluded:
        variables_shared.append(var)
    
    print('///////////////////////////////////////////////////////////////////////////////')
    print('Shared Variables to Update:')
    for var in variables_shared:
        print(local_rank, ': ',  var.op.name, var)
    print('///////////////////////////////////////////////////////////////////////////////')
    print('Unshared Variables to Update:')
    for var in variables_not_shared:
        print(local_rank, ': ', var.op.name, var)
    print('///////////////////////////////////////////////////////////////////////////////')

    opt_distributed = hvd.DistributedOptimizer(opt)
    print('hvd.size: ', hvd.size())
    global_step = tf.train.get_or_create_global_step()
    # Make training operation
    #train_op = opt_distributed.minimize(total_loss, global_step=global_step)
    **grads_vars_shared = opt_distributed.compute_gradients(total_loss, var_list=variables_shared)
    for g,v in grads_vars_shared:
        print(g,v)
    grads_vars_not_shared = opt.compute_gradients(total_loss, var_list=variables_not_shared)
    for g,v in grads_vars_not_shared:
        print(g,v)**

    #train_op_shared = opt_distributed.apply_gradients(grads_vars_shared, global_step=global_step)
    #train_op_not_shared = opt.apply_gradients(grads_vars_not_shared, global_step=global_step)
    

    subdir = datetime.strftime(datetime.now(), '%Y%m%d-%H%M%S')
    # Save checkpoints only on worker 0 to prevent other workers from corrupting them.
    checkpoint_dir = os.path.join('temp', subdir) if hvd.rank() == 0 else None
    print('Model Directory:', checkpoint_dir)

    feed_dict_1 = {files_placeholder_1: files_set_1, labels_placeholder_1: labels_set_1}
    feed_dict_2 = {files_placeholder_2: files_set_2, labels_placeholder_2: labels_set_2}
    initializer_hook_1 = _DatasetInitializerHook(init_iterator_1, feed_dict_1)
    initializer_hook_2 = _DatasetInitializerHook(init_iterator_2, feed_dict_2)

    if local_rank == 0:
        initializer_hook = initializer_hook_1
    else:
        initializer_hook = initializer_hook_2
        
    # Add hook to broadcast variables from rank 0 to all other processes during
    # initialization.
    hooks = [
        hvd.BroadcastGlobalVariablesHook(0),
        initializer_hook,

        # Horovod: adjust number of steps based on number of GPUs.
        tf.train.StopAtStepHook(last_step=10000000 // hvd.size()),

        tf.train.LoggingTensorHook(tensors={'step': global_step, \
                                            'total_loss': total_loss, \
                                            'cross_entropy': cross_entropy_mean, 'accuracy': accuracy, \
                                            'learning_rate': lr_placeholder \
                                              }, every_n_iter=100),
    ]

    with tf.train.MonitoredTrainingSession(checkpoint_dir=checkpoint_dir,
                                           config=config,
                                           save_checkpoint_steps=20000,
                                           hooks=hooks) as mon_sess:
        steps = []
        lrs = []
        if args.learning_rate < 0:
            step_lrs = args.learning_rate_multistep.split()
            for ele in step_lrs:
                ele = ele.split(',')
                steps.append(int(ele[0]))
                lrs.append(float(ele[-1]))
            print('learning rate based on multistep value: ', steps, lrs)

        step = 0
        while not mon_sess.should_stop():
            if args.learning_rate > 0.0:
                lr = args.learning_rate
            else:
                lr = get_learning_rate_step(lrs, steps, step)

            # Perform synchronous training.
            mon_sess.run([grads_vars_shared], feed_dict={lr_placeholder:lr})
            step += 1

LOG Information：

I0117 08:19:00.281375 139968497534784 session_manager.py:500] Running local_init_op.
I0117 08:19:00.281362 139721998829376 session_manager.py:500] Running local_init_op.
I0117 08:19:00.347552 139968497534784 session_manager.py:502] Done running local_init_op.
I0117 08:19:00.347606 139721998829376 session_manager.py:502] Done running local_init_op.
[JXQ-240-26-97.h.chinabank.com.cn:27709] Read -1, expected 6273, errno = 1
I0117 08:19:03.174535 139721998829376 basic_session_run_hooks.py:606] Saving checkpoints for 0 into temp/20200117-081854/model.ckpt.
[2020-01-17 08:20:34. 24110: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ranks:
0: [HorovodBroadcast_acm_vggface2_kernel_0]
1: [HorovodBroadcast_acm_webface_kernel_0]
[2020-01-17 08:21:34[1,0]<stderr>:. 27200: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ranks:
0: [HorovodBroadcast_acm_vggface2_kernel_0]
1: [HorovodBroadcast_acm_webface_kernel_0]
[2020-01-17 08:22:34. 28136: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ranks:
0: [HorovodBroadcast_acm_vggface2_kernel_0]
1: [HorovodBroadcast_acm_webface_kernel_0]
[2020-01-17 08:23:34. 30789: W horovod/common/operations.cc:779] One or more tensors were submitted to be reduced, gathered or broadcasted by subset of ranks and are waiting for remainder of ranks for more than 60 seconds. This may indicate that different ranks are trying to submit different tensors or that only subset of ranks is submitting tensors, which will cause deadlock. 
Stalled ranks:
0: [HorovodBroadcast_acm_vggface2_kernel_0]
1: [HorovodBroadcast_acm_webface_kernel_0]

---------------------------------------------------------------------------
acm_vggface2_kernel is the name of the weight of fc layer.

**Comment by gobigrassland** [comment]
test
